lib/Dialect/Poly/PolyOps.cpp

```diff
@@ -8,6 +11,51 @@ OpFoldResult ConstantOp::fold(ConstantOp::FoldAdaptor adaptor) {
   return adaptor.getCoefficients();
 }
 
-}  // namespace poly
-}  // namespace tutorial
-}  // namespace mlir
+OpFoldResult AddOp::fold(AddOp::FoldAdaptor adaptor) {
+  return constFoldBinaryOp<IntegerAttr, APInt>(
+      adaptor.getOperands(), [&](APInt a, APInt b) { return a + b; });
+}
+
+OpFoldResult SubOp::fold(SubOp::FoldAdaptor adaptor) {
+  return constFoldBinaryOp<IntegerAttr, APInt>(
```

Comment:
The calling of `constFoldBinaryOp<IntegerAttr, APInt>` changes in recent versions to `constFoldBinaryOp<IntegerAttr, APInt, void>` as per [here](https://github.com/llvm/llvm-project/blob/llvmorg-18.1.5/mlir/include/mlir/Dialect/CommonFolders.h#L162) following the introduction of the `PoisonAttr` support via the `ub` dialect in [this commit](https://github.com/llvm/llvm-project/commit/5dce74817b71a1f646fb2857c037b3a66f41c7cd). 

Otherwise, by default, the template matches to `constFoldBinaryOp<IntegerAttr, APInt, ub::PoisonAttr>` which requires us to add a dependency to the `ub` dialect. I haven't quite figured out how to do that yet but setting the 3rd parameter to `void` preserves the old behaviour to the extent of my testing.

Reply:
I noticed the same thing for different reasons, but this is part of why I have the LLVM version dependency pinned to a specific commit in the bazel workspace configuration.

Still, a good note for others encountering the same issue. The APIs will drift over time. I am working on upstreaming this tutorial to the main LLVM repo, so that these kinds of issues can be systematically addressed, but until then we'll have to settle for staleness :)